### ISSUE TITLE
kubernetes_secret provider resource fails with v1.3.X? versions of the api

### ISSUE THREAD

**Issue by dkiser** [hashicorp/terraform#14227]
### Terraform Version
`Terraform v0.9.4`

### Affected Resource(s)
- kubernetes_secret

### Terraform Configuration Files

> Some portion redacted, but idea is there.

```hcl
 data "template_file" "dockerconfigjson" {
     count = "${REDACTED.cstf.count}"

     template = "${file("templates/dockerconfigjson.tpl")}"
     vars {
         # repo name is of the format "docker-<effort_name>-repo-REDACTED"
         repo_name = "docker-${element(kubernetes_namespace.cstf.*.metadata.0.name, count.index)}-repo-REDACTED"
         # base64 encode the "user:password" for auth_token
         # here we will use the repouser user/pass again
         auth_token = "${base64encode("${element(data.REDACTED_repouser.cstf.*.name, count.index)}:${element(data.REDACTED_repouser.cstf.*.password, count.index)}")}"
         email = "REDACTED"
     }
 }

 resource "kubernetes_secret" "k8s-artifactory-creds" {
     count = "${REDACTED.cstf.count}"

     metadata {
         namespace = "${element(kubernetes_namespace.REDACTED.*.metadata.0.name, count.index)}"
         name = "k8s-artifactory-creds"
     }

     data {
         #base64 encode entire dockerconfigjson
         ".dockerconfigjson" = "${base64encode("${element(data.template_file.dockerconfigjson.*.rendered, count.index)}")}"
     }

     type = "kubernetes.io/dockerconfigjson"
 }
```

### Debug Output
https://gist.github.com/dkiser/6dc4e6757504ebbab3c0289c7e84a64e

### Panic Output
N/A

### Expected Behavior
Expected behavior on kubernetes API v1.3.5 is to create the secret.

### Actual Behavior
What actually happened?

**error from k8s api:**
`* kubernetes_secret.k8s-artifactory-creds: Secret "k8s-artifactory-creds" is invalid: data[.dockerconfigjson]: Required value`

### Steps to Reproduce

1. Configure a terraform provider for kubernetes 1.3.5
2. Create resources similar to shown above
3. `terraform apply`

### Important Factoids
Looking at k8s source code for [v1.6](https://github.com/kubernetes/kubernetes/blob/release-1.6/pkg/api/v1/types.go#L4066) vs [1.3](https://github.com/kubernetes/kubernetes/blob/release-1.3/pkg/api/v1/types.go#L3090) it appears that `type Secret` varies.  The issue here is possibly  the kubernetes provider in terraform is coded against a newer go gettable version of `"k8s.io/kubernetes/pkg/api/v1"`, which contains the `StringData` field in the `type Secret` struct.  This seems to be filled out correctly in terraform, but the k8s server expects all data to be in the `Data` field of the `type Secret` struct as its serialized to json consumption by the API (presumably because it was compiled against older versions of the `pkg/api/v1` in the k8s source tree).

I wonder if there is a possibility to detect versions of the server the kubernetes provider is talking to and adjust accordingly.

The implemented provider is always using the `StringData` field per https://github.com/hashicorp/terraform/blob/master/builtin/providers/kubernetes/resource_kubernetes_secret.go#L50

### References
N/A

**Comment by radeksimko** [comment]
Hi @dkiser 
sorry for the inconvenience.

The initial implementation of the K8S provider is compiled with vendored K8S `1.5.3`. The backward compatibility **from 1.5.3 on** will be maintained on best effort basis.

Vendoring K8S means vendoring ~500k+ LOC for us, which is a lot 😅 , vendoring multiple versions would mean N times 500k (and we probably couldn't do it anyway due to strict 1-1 mapping between import path & dependency) which is why I'm not too inclined to support multiple versions in a single Terraform release.

Once we bump K8S API/SDK to `1.6.x` we'll retest all resources and see if we need to make any changes to any schema and/or add state migrations, so that such upgrade is as smooth as possible. Once we release such change, any upcoming versions will be 1.6.x compatible (with whatever backward compatibility K8S API/SDK offers under the hood).

For the above reasons I don't personally plan to spend time on this, but if you know how to make it compatible without changing `/vendor` I'm happy to review a PR.

TL;DR It may be easiest to just upgrade to `1.5.x`

I relabelled this issue as we should mention the compatible K8S version(s) in the docs. I plan to close this once we update docs - which shouldn't prevent anyone from submitting a PR though.

Sorry for the bad news.

**Comment by dkiser** [comment]
@radeksimko Thanks for the response!  It was exactly the response I was expecting 8-).  Completely understand.  

A quick workaround was to use a null resource with local provisioner and/or adding a local provisioner to the namespace resource.
